design/jwt-verification-design.md

```diff
@@ -0,0 +1,205 @@
+# JWT Verification Support
+
+## Abstract
+This document describes a design for performing JSON Web Token (JWT) verification for requests to virtual hosts hosted by Contour.
+
+## Background
+JSON Web Token (JWT) is an open standard (RFC 7519) that defines a compact and self-contained way for securely transmitting information between parties as a JSON object. 
+This information can be verified and trusted because it is digitally signed. (ref. https://jwt.io/introduction)
+
+Envoy Proxy has built-in support for verifying JWTs that are attached to incoming requests, via the [JWT Authentication HTTP filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/jwt_authn_filter#config-http-filters-jwt-authn). (Note, this document will use the term **JWT verification** to describe this process).
+Specifically, Envoy can verify the signature, audience, issuer and time restrictions of a JWT.
+If verification fails, the request will be rejected.
+It's important to note that this filter does not itself obtain a JWT for an incoming request; the request must already have one attached.
+
+Contour does not currently have support for configuring the JWT authentication filter.
+This document proposes a design for adding that support to Contour's custom resource, `HTTPProxy`.
+
+## Goals
+- JWT verification for requests to TLS-enabled virtual hosts.
+- Expose a subset of the Envoy filter configuration to cover the most common use cases.
+- Be able to easily expose additional configuration settings if/when needed in the future.
+
+## Non Goals
+- JWT verification for requests to non-TLS enabled virtual hosts.
+- Exposing all possible Envoy configuration settings.
+- Supporting end-to-end OAuth2/OIDC flows.
+
+## High-Level Design
+Contour's `HTTPProxy` resource will get a new optional field, `spec.virtualhost.jwtProviders`, to define the details of how to verify JWTs for requests to a given virtual host.
+This field will only be supported for virtual hosts for which Envoy is terminating TLS.
```

Comment:
Mostly because there is only one filter chain for all plain HTTP virtualhosts, so only one filter instance and no "straightforward" way to configure different providers for different vhosts

(if we want to offer this, maybe we could combine the providers for different vhosts all together in the same filter configuration, but make their names unique etc.)

Reply:
Yeah it's really for both of those reasons. I do think we _could_ make it work for plain HTTP -- if we add an `:authority` header match for each requirement, then we should be able to scope things for the right vhosts - but we'd have to be careful to ensure no pollution across vhosts. My preference would be to not implement it unless there's a strong use case for it.